### PR TITLE
Add BC promise rules for constructors

### PR DESCRIPTION
**Pull request by alexandre-daubois**
As spotted in https://github.com/symfony/symfony/pull/58490, following https://github.com/symfony/symfony/pull/58490#discussion_r1791885700

### CODE REVIEW

**Review by OskarStark — APPROVED** [review]

File: contributing/code/bc.rst [hunk 1]
```diff
@@ -258,6 +258,14 @@ Make public or protected                                                  Yes
 Remove private property                                                   Yes
 **Constructors**
 Add constructor without mandatory arguments                               Yes             :ref:`[1] <note-1>`
+:ref:`Add argument without a default value <add-argument-public-method>`  No
+Add argument with a default value                                         Yes
```
Comment (wouterj) [inline contributing/code/bc.rst]: We should add a footnote that this is only allowed at the end of the parameter list (similar to `[3]`, but then replacing removing with adding).

File: contributing/code/bc.rst [hunk 2]
```diff
@@ -258,6 +258,14 @@ Make public or protected                                                  Yes
...
+Remove argument                                                           No              :ref:`[3] <note-3>`
+Add default value to an argument                                          Yes
+Remove default value of an argument                                       No
+Add type hint to an argument                                              No
+Remove type hint of an argument                                           No
+Change argument type                                                      No
```
Comment (wouterj) [inline contributing/code/bc.rst]: These 3 are allowed, if the new type is compatible with the previous type (e.g. you can change from `SomeObject` to `SomeObject|OrAnotherOne`, but not to `string`).

For public final methods, we treat this as a plain "Yes", as the "compatible with existing state" is already covered by the "Using our Classes" promise. Let's do the same here.

**Review by wouterj — COMMENTED** [review]
Thanks! We should move this new rows at the end of the `**Constructors**` section I think, to be consistent with the methods section.

File: contributing/code/bc.rst [hunk 3]
```diff
@@ -258,6 +258,14 @@ Make public or protected                                                  Yes
 Remove private property                                                   Yes
 **Constructors**
 Add constructor without mandatory arguments                               Yes             :ref:`[1] <note-1>`
+:ref:`Add argument without a default value <add-argument-public-method>`  No
+Add argument with a default value                                         Yes
+Remove argument                                                           No              :ref:`[3] <note-3>`
+Add default value to an argument                                          Yes
+Remove default value of an argument                                       No
+Add type hint to an argument                                              No
+Remove type hint of an argument                                           No
```
Comment (stof) [inline contributing/code/bc.rst]: this is actually allowed on a constructor, as it is a type of changed allowed for argument types (allowing anything is a super type of any existing type)

File: contributing/code/bc.rst [hunk 4]
```diff
@@ -258,6 +258,14 @@ Make public or protected                                                  Yes
 Remove private property                                                   Yes
 **Constructors**
 Add constructor without mandatory arguments                               Yes             :ref:`[1] <note-1>`
+:ref:`Add argument without a default value <add-argument-public-method>`  No
+Add argument with a default value                                         Yes
+Remove argument                                                           No              :ref:`[3] <note-3>`
+Add default value to an argument                                          Yes
+Remove default value of an argument                                       No
+Add type hint to an argument                                              No
```
Comment (stof) [inline contributing/code/bc.rst:266]: we should probably update the whole table (maybe in a separate PR) to stop talking about `type hint`. Those are not hints at all. the right term for that is `type`, or maybe `native type` if we want to compare it to phpdoc-declared types.

File: contributing/code/bc.rst (hunk 2 above)
Comment (stof) [reply contributing/code/bc.rst]: why would public final methods use a plain `Yes` ? The restriction is not about inheritance here, so it is not different for final constructors.

File: contributing/code/bc.rst (hunk 4 above)
Comment (alexandre-daubois) [reply contributing/code/bc.rst:266]: I agree, I'll have a loot at it once this one is merged

File: contributing/code/bc.rst (hunk 3 above)
Comment (alexandre-daubois) [reply contributing/code/bc.rst]: That's right, updated

File: contributing/code/bc.rst (hunk 1 above)
Comment (alexandre-daubois) [reply contributing/code/bc.rst]: New footnote added, does it look good to you?

**Comment by javiereguiluz** [comment]
Merged! Thanks Alex.

If further tweaks are needed, we can do that in future PRs. Thanks.

**Pull request merged**